ocaml/xapi/repository.ml

```diff
@@ -114,3 +129,476 @@ let cleanup_pool_repo () =
+      | `Succ s -> `Ok s
+      | `Fail e -> `Error e
+    in
+    let to_be_run, remaining = split_for_first_n [] capacity funs in
+    match to_be_run with
+    | [] -> acc
+    | _ ->
+      let finished =
+        List.map run to_be_run
+        |> List.map get_result
+        |> List.map (function `Ok s -> s | `Error e -> raise e)
+      in
+      run_in_parallel' (List.rev_append finished acc) remaining capacity
+  in
+  run_in_parallel' [] funs capacity
+
+let http_get_host_updates_in_json ~__context ~host ~installed =
+  let host_session_id =
+    Xapi_session.login_no_password ~__context ~uname:None ~host ~pool:true
+      ~is_local_superuser:true ~subject:Ref.null ~auth_user_sid:""
+      ~auth_user_name:"" ~rbac_permissions:[]
+  in
+  let request = Xapi_http.http_request
+      ~cookie:[("session_id", Ref.string_of host_session_id)]
+      ~query:[("installed", (string_of_bool installed))]
+      Http.Get Constants.get_host_updates_uri
+  in
+  let host_name = (Db.Host.get_hostname ~__context ~self:host) in
+  let host_addr = Db.Host.get_address ~__context ~self:host in
+  let open Xmlrpc_client in
+  let transport = SSL (SSL.make () ~verify_cert:false, host_addr, !Constants.https_port) in
+  debug "getting host updates on %s (addr %s) by HTTP GET" host_name host_addr;
+  Xapi_stdext_pervasives.Pervasiveext.finally
+    (fun () ->
+      try
+        let json_str =
+          with_transport transport
+            (with_http request (fun (response, fd) ->
+                 Xapi_stdext_unix.Unixext.string_of_fd fd))
+        in
+        debug "host %s returned updates: %s" host_name json_str;
+        Yojson.Basic.from_string json_str
+      with e ->
+        let uuid = Db.Host.get_uuid ~__context ~self:host in
+        error "Failed to get updates from host uuid='%s': %s" uuid (ExnHelper.string_of_exn e);
+        raise Api_errors.(Server_error (get_host_updates_failed, [uuid])))
+    (fun () -> Xapi_session.destroy_db_session ~__context ~self:host_session_id)
+
+let set_available_updates ~__context ~self =
+  let hosts = Db.Host.get_all ~__context in
+  let xml_path =
+    "repodata/repomd.xml"
+    |> Filename.concat !Xapi_globs.pool_repo_name
+    |> Filename.concat !Xapi_globs.local_pool_repo_dir
+  in
+  let md = parse_repomd xml_path in
+  let are_updates_available host () =
+    let json = http_get_host_updates_in_json ~__context ~host ~installed:false in
+    (* TODO: live patches *)
+    match Yojson.Basic.Util.member "updates" json with
+    | `List [] -> false
+    | _ -> true
+    | exception e ->
+      let uuid = Db.Host.get_uuid ~__context ~self:host in
+      error "Invalid updates from host uuid='%s': %s" uuid (ExnHelper.string_of_exn e);
+      raise Api_errors.(Server_error (get_host_updates_failed, [uuid]))
+  in
+  let funs = List.map (fun h -> are_updates_available h) hosts in
+  let rets_of_hosts = run_in_parallel funs capacity_in_parallel in
```

Comment:
If `are_updates_available` fails on any of the hosts we'll throw `get_host_updates_failed` which I believe is correct

Reply:
Yes, well done on this quite complicated procedure. It looks right to me.

Reply:
Once joined all hosts in one batch, and if there is an error, it will raise exception, and will not continue on next batch.
But in one batch, I think it's better to wait for joining all (e.g. 16) hosts. Otherwise the main thread may lose control on un-finished threads.